### PR TITLE
VDS-1750 Implement an backend for the customers export

### PR DESCRIPTION
**Pull request by trueboroda**
Description: 

--

QA-test: 

Demo-test: 

Download artifact URL: https://vc3prerelease.blob.core.windows.net/packages/VirtoCommerce.CustomerExportImport_1.0.0-pr-13-870fa870.zip

### CODE REVIEW

**Comment by mvktsk** [comment]
Review task https://virtocommerce.atlassian.net/browse/VP-7419 has been created

File: src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportPushNotification.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        {
+        }
+
+        public string JobId { get; set; }
+
+        public int ProcessedCount { get; set; }
+
+        public int TotalCount { get; set; }
+
+        public ICollection<string> Errors { get; set; }
```
Comment (pushnitsa) [inline src/VirtoCommerce.CustomerExportImportModule.Core/Models/ExportPushNotification.cs:20]: Initialize collection in the constructor

File: src/VirtoCommerce.CustomerExportImportModule.Core/Services/IExportWriter.cs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+using System;
+
+namespace VirtoCommerce.CustomerExportImportModule.Core.Services
+{
+    public interface IExportWriter<in T> : IDisposable
+    {
+        public void WriteRecords(T[] records);
```
Comment (pushnitsa) [inline src/VirtoCommerce.CustomerExportImportModule.Core/Services/IExportWriter.cs:7]: public here is redundant

File: src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            var organizationFilePath = GetExportFilePath("Organizations");
+            var organizationExportWriter = _exportWriterFactory.Create<ExportableOrganization>(organizationFilePath, new ExportConfiguration());
+
+            try
+            {
+                while (await dataSource.FetchAsync())
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var contacts = dataSource.Items.Select(x => x as ExportableContact).Where(x => x != null).ToArray();
```
Comment (pushnitsa) [inline src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs:56]: Maybe will be better to use `Cast<T>` LINQ method?

File: src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                contactExportWriter.Dispose();
+                organizationExportWriter.Dispose();
+            }
+        }
+
+        private string GetExportFilePath(string entityName)
+        {
+            if (string.IsNullOrEmpty(_platformOptions.DefaultExportFolder))
+            {
+                throw new PlatformException($"{nameof(_platformOptions.DefaultExportFolder)} must be set.");
```
Comment (pushnitsa) [inline src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs:85]: Platform exception? Maybe ArgumentNullException?

File: src/VirtoCommerce.CustomerExportImportModule.Data/Services/ExportWriter.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+using CsvHelper.Configuration;
+using VirtoCommerce.CustomerExportImportModule.Core.Services;
+using VirtoCommerce.CustomerExportImportModule.Data.ExportImport;
+using VirtoCommerce.Platform.Core.Assets;
+
+namespace VirtoCommerce.CustomerExportImportModule.Data.Services
+{
+    public sealed class ExportWriter<T> : IExportWriter<T>
+    {
+
```
Comment (pushnitsa) [inline src/VirtoCommerce.CustomerExportImportModule.Data/Services/ExportWriter.cs:12]: Remove empty line

File: src/VirtoCommerce.CustomerExportImportModule.Web/BackgroundJobs/ExportJob.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                await _customerDataExporter.ExportAsync(request,
+                    progressInfo => ProgressCallback(progressInfo, pushNotification, context),
+                    new JobCancellationTokenWrapper(jobCancellationToken));
+            }
+            catch (JobAbortedException)
+            {
+                // job is aborted, do nothing
+            }
+
```
Comment (pushnitsa) [inline src/VirtoCommerce.CustomerExportImportModule.Web/BackgroundJobs/ExportJob.cs:39]: Remove empty line

File: src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs (hunk 3 above)
Comment (pushnitsa) [reply src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs:56]: Or `OfType<T>`

File: src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs (hunk 4 above)
Comment (trueboroda) [reply src/VirtoCommerce.CustomerExportImportModule.Data/Services/CustomerDataExporter.cs:85]: It is a platform option. It is not the argument.

**Review by pushnitsa — APPROVED** [review]

**Pull request merged**